packages/apollo-engine-reporting/src/contextualizedStats.ts

```diff
@@ -0,0 +1,246 @@
+}
+
+export class FieldStat implements IFieldStat {
+  returnType: string;
+  errorsCount: number = 0;
+  count: number = 0;
+  requestsWithErrorsCount: number = 0;
+  latencyCount: DurationHistogram = new DurationHistogram();
+
+  constructor(returnType: string) {
+    this.returnType = returnType;
+  }
+}
+
+export class ContextualizedStats {
+  statsContext: IStatsContext;
+  queryLatencyStats: QueryLatencyStats;
+  perTypeStat: { [k: string]: TypeStat };
+
+  constructor(statsContext: IStatsContext) {
+    this.statsContext = statsContext;
+    this.queryLatencyStats = new QueryLatencyStats();
+    this.perTypeStat = Object.create(null);
+  }
+
+  public addTrace(trace: Trace) {
+    const queryLatencyStats = this.queryLatencyStats;
+    queryLatencyStats.requestCount++;
+    if (trace.fullQueryCacheHit) {
+      queryLatencyStats.cacheLatencyCount.incrementDuration(trace.durationNs);
+      queryLatencyStats.cacheHits++;
+    } else {
+      queryLatencyStats.latencyCount.incrementDuration(trace.durationNs);
+    }
+
+    if (
+      !trace.fullQueryCacheHit &&
+      trace.cachePolicy &&
+      trace.cachePolicy.maxAgeNs
+    ) {
+      if (trace.cachePolicy.scope == Trace.CachePolicy.Scope.PRIVATE) {
+        queryLatencyStats.privateCacheTtlCount.incrementDuration(
+          trace.cachePolicy.maxAgeNs,
+        );
+      } else if (trace.cachePolicy.scope == Trace.CachePolicy.Scope.PUBLIC) {
+        queryLatencyStats.publicCacheTtlCount.incrementDuration(
+          trace.cachePolicy.maxAgeNs,
+        );
+      }
+    }
+
+    if (trace.persistedQueryHit) {
+      queryLatencyStats.persistedQueryHits++;
+    } else if (trace.persistedQueryRegister) {
+      queryLatencyStats.persistedQueryMisses++;
+    }
+
+    if (trace.forbiddenOperation) {
+      queryLatencyStats.forbiddenOperationCount++;
+    } else if (trace.registeredOperation) {
+      queryLatencyStats.registeredOperationCount++;
+    }
+
+    let hasError = false;
+    const typeStats = this.perTypeStat;
+    const rootPathErrorStats = queryLatencyStats.rootErrorStats;
+
+    function traceNodeStats(node: Trace.INode, path: ReadonlyArray<string>): boolean {
+      // Generate error stats and error path information
+      if (node.error && node.error.length > 0) {
+        hasError = true;
+
+        let currPathErrorStats = rootPathErrorStats;
+
+        for (const subPathEntry of path.entries()) {
+          // Using entries instead values since Node 8
+          // doesn't support Array.prototype.values
+          const subPath = subPathEntry[1];
+          let children =
+            currPathErrorStats.children ||
+            (currPathErrorStats.children = Object.create(null));
+          currPathErrorStats =
+            children[subPath] || (children[subPath] = Object.create(null));
+        }
+
+        currPathErrorStats.requestsWithErrorsCount =
+          (currPathErrorStats.requestsWithErrorsCount || 0) + 1;
```

Comment:
I think this would be more clear if you used a class instead of Object.create(null) here, so that you got 0 defaults. In fact the codegen exports a PathErrorStats class that's fine I think? You need to make your own classes above so that you can swap in DurationHistogram and maps and stuff, but for this you can just use the codegenned class.